### ISSUE TITLE
Issue with thumbnails on S3 image upload

### ISSUE THREAD

**Issue by swisscat** [servocoder/RichFilemanager-PHP#19]
Attempting to upload an image causes the following issues:

1. Similar to issue https://github.com/servocoder/RichFilemanager-PHP/issues/18, a directory for thumbnails is attempted to be created. Again fixed on my side using https://github.com/swisscat/RichFilemanager-PHP/commit/f4944f74cebb5a66d7a1fd592dba061fffa1edb1, but I think this definitely needs a broader approach ;)

1. Post that, I get the two following issues:

```
[25-Oct-2017 15:10:52 Europe/Berlin] PHP Warning:  Error executing "PutObject" on "https://<bucket>.s3-eu-west-1.amazonaws.com/var/06e741ad-5ed3-4022-ac7d-e4dd2cd8bccd.jpg"; AWS HTTP error: Client error: `PUT https://<bucket>.s3-eu-west-1.amazonaws.com/var/06e741ad-5ed3-4022-ac7d-e4dd2cd8bccd.jpg` resulted in a `403 Forbidden` response:
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>Access Denied</Message><RequestId>F66E86 (truncated...)
 AccessDenied (client): Access Denied - <?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>Access Denied</Message><RequestId>F66E86D8BC0AD339</RequestId><HostId>63rOdlyHgx+hxfEUB8wV1CxToc37s1N0cGQUA9aRPOtdoEtQITE3fhYbbprObVV3OdZ0PpvakUQ=</HostId></Error> in /var/www/html/vendor/aws/aws-sdk-php/src/S3/StreamWrapper.php on line 738
[25-Oct-2017 15:10:52 Europe/Berlin] PHP Stack trace:
[25-Oct-2017 15:10:52 Europe/Berlin] PHP   1. {main}() /var/www/html/vendor/servocoder/richfilemanager/connectors/php/filemanager.php:0
[25-Oct-2017 15:10:52 Europe/Berlin] PHP   2. RFM\Application->run() /var/www/html/vendor/servocoder/richfilemanager/connectors/php/filemanager.php:70
[25-Oct-2017 15:10:52 Europe/Berlin] PHP   3. RFM\Api\AwsS3Api->actionUpload() /var/www/html/vendor/servocoder/richfilemanager-php/src/Application.php:314
[25-Oct-2017 15:10:52 Europe/Berlin] PHP   4. RFM\Repository\S3\UploadHandler->post($print_response = *uninitialized*) /var/www/html/vendor/servocoder/richfilemanager-php/src/Api/AwsS3Api.php:133
[25-Oct-2017 15:10:52 Europe/Berlin] PHP   5. RFM\Repository\S3\UploadHandler->handle_file_upload($uploaded_file = *uninitialized*, $name = *uninitialized*, $size = *uninitialized*, $type = *uninitialized*, $error = *uninitialized*, $index = *uninitialized*, $content_range = *uninitialized*) /var/www/html/vendor/servocoder/richfilemanager-php/src/Repository/BaseUploadHandler.php:1449
[25-Oct-2017 15:10:52 Europe/Berlin] PHP   6. RFM\Repository\S3\UploadHandler->handle_image_file($file = *uninitialized*) /var/www/html/vendor/servocoder/richfilemanager-php/src/Repository/S3/UploadHandler.php:279
[25-Oct-2017 15:10:52 Europe/Berlin] PHP   7. RFM\Repository\S3\UploadHandler->create_scaled_image($file = *uninitialized*, $version = *uninitialized*, $options = *uninitialized*) /var/www/html/vendor/servocoder/richfilemanager-php/src/Repository/BaseUploadHandler.php:1074
[25-Oct-2017 15:10:52 Europe/Berlin] PHP   8. RFM\Repository\S3\UploadHandler->gd_create_scaled_image($file = *uninitialized*, $version = *uninitialized*, $options = *uninitialized*) /var/www/html/vendor/servocoder/richfilemanager-php/src/Repository/BaseUploadHandler.php:1039
[25-Oct-2017 15:10:52 Europe/Berlin] PHP   9. imagejpeg(*uninitialized*, *uninitialized*, *uninitialized*) /var/www/html/vendor/servocoder/richfilemanager-php/src/Repository/BaseUploadHandler.php:811
[25-Oct-2017 15:10:52 Europe/Berlin] PHP  10. Aws\S3\StreamWrapper->stream_flush() /var/www/html/vendor/servocoder/richfilemanager-php/src/Repository/BaseUploadHandler.php:811
[25-Oct-2017 15:10:52 Europe/Berlin] PHP  11. Aws\S3\StreamWrapper->boolCall($fn = *uninitialized*, $flags = *uninitialized*) /var/www/html/vendor/aws/aws-sdk-php/src/S3/StreamWrapper.php:180
[25-Oct-2017 15:10:52 Europe/Berlin] PHP  12. Aws\S3\StreamWrapper->triggerError($errors = *uninitialized*, $flags = *uninitialized*) /var/www/html/vendor/aws/aws-sdk-php/src/S3/StreamWrapper.php:912
[25-Oct-2017 15:10:52 Europe/Berlin] PHP  13. trigger_error(*uninitialized*, *uninitialized*) /var/www/html/vendor/aws/aws-sdk-php/src/S3/StreamWrapper.php:738
[25-Oct-2017 15:10:52 Europe/Berlin] PHP Fatal error:  Uncaught InvalidArgumentException: Found 1 error while validating the input provided for the PutObject operation:
[Key] is missing and is a required parameter in /var/www/html/vendor/aws/aws-sdk-php/src/Api/Validator.php:65
Stack trace:
#0 /var/www/html/vendor/aws/aws-sdk-php/src/Middleware.php(79): Aws\Api\Validator->validate('PutObject', Object(Aws\Api\StructureShape), Array)
#1 /var/www/html/vendor/aws/aws-sdk-php/src/S3/S3Client.php(421): Aws\Middleware::Aws\{closure}(Object(Aws\Command), NULL)
#2 /var/www/html/vendor/aws/aws-sdk-php/src/S3/S3Client.php(444): Aws\S3\S3Client::Aws\S3\{closure}(Object(Aws\Command), NULL)
#3 /var/www/html/vendor/aws/aws-sdk-php/src/S3/S3Client.php(378): Aws\S3\S3Client::Aws\S3\{closure}(Object(Aws\Command), NULL)
#4 /var/www/html/vendor/aws/aws-sdk-php/src/S3/S3Client.php(397): Aws\S3\S3Client::Aws\S3\{closure}(Object(Aws\Command), NULL)
#5 /var/www/html/vendor/aws/aws-sdk-php/src/Middleware.php(55): Aws\S3\S3Client::Aws\S3\{closure}(Object(Aws\Command), NULL)
#6 /var/www/ht in /var/www/html/vendor/aws/aws-sdk-php/src/Api/Validator.php on line 65
```

**Comment by servocoder** [comment]
Thanks for the reporting.
I have an idea how to fix it globally. I will provide the fix soon.

**Comment by servocoder** [comment]
Try this commit, and let me know if it solve the problem: https://github.com/servocoder/RichFilemanager-PHP/commit/49de88e5cb7c14097eb81ce040780ea4bf86e277

**Comment by swisscat** [comment]
This looks better, however, still an issue with the thumbnails creation, due to lack of SSE flag passed (happens when called from https://github.com/servocoder/RichFilemanager-PHP/blob/master/src/Repository/BaseUploadHandler.php#L706)

I see the issue, however, I wonder what's the best course of action:

1. Modify https://github.com/servocoder/RichFilemanager-PHP/blob/master/src/Repository/S3/StorageHelper.php#L108 to register the stream wrapper taking server-side encryption as a context option (which might be too generic to do so? I haven't entirely figured out how to do so)

1. Put a layer of abstraction to avoid calling these low-level functions

**Comment by servocoder** [comment]
I see the problem. Thanks for your ideas.
I will investigate it deeply and get back with a solution.

**Comment by servocoder** [comment]
I have included encryption type as default stream wrapper parameter and also added it to direct S3Client functions (GetObject, PutObject, ListObjects). All encrypted read-write operations should work now.

I included it into v1.1.2 that has been published. I close the issue but I would be grateful for your feedback. HAve a good day.

**Comment by swisscat** [comment]
@servocoder , version 1.1.3 is fixing all the issues. Thanks for the intervention!
